tools/pre_ci.py

```
@@ -403,12 +404,29 @@ def execute(self) -> None:
             self.run_commands_parallel(parallel_tasks)
 
             # 4. Unit Tests (Run sequentially to see clear progress)
-            # Pytest executes in the remote test-matrix job. Prevent redundancy in CI.
             if not self.is_ci:
-                self.run_command(
+                print("\n>>> [Step: Unit Tests & Coverage Enforcement]", flush=True)
+                # Stream output directly to console for real-time visibility.
+                # Coverage failure is detected via pytest's own exit code when
+                # --cov-fail-under is passed correctly in pyproject.toml.
+                res = subprocess.run(
                     ["uv", "run", "--no-sync", "pytest", "-v"],
```

Ruff (0.15.12)

[error] Starting a process with a partial executable path

(S607)